builtin/common/serialize.lua

```diff
@@ -1,205 +1,217 @@
-		if nv then seen[v] = 2 end
-	end
-
-	-- First phase, list the tables and functions which appear more than
-	-- once in x.
-	local function mark_multiple_occurences(x)
-		local tp = type(x)
-		if tp ~= "table" and tp ~= "function" then
-			-- No identity (comparison is done by value, not by instance)
-			return
-		end
-		if seen[x] == 1 then
-			seen[x] = 2
-		elseif seen[x] ~= 2 then
-			seen[x] = 1
-		end
+local table_insert, table_concat, string_dump, string_format, string_match, math_huge
+	= table.insert, table.concat, string.dump, string.format, string.match, math.huge
 
-		if tp == "table" then
-			nested[x] = true
-			for k, v in pairs(x) do
-				if nested[k] or nested[v] then
-					mark_nest_point(x, k, v)
-				else
-					mark_multiple_occurences(k)
-					mark_multiple_occurences(v)
-				end
+-- Recursively counts occurences of objects (non-primitives including strings) in a table.
+local function count_objects(value)
+	local counts = {}
+	if value == nil then
+		-- Early return for nil
+		return counts
+	end
+	local function count_values(val)
+		local type_ = type(val)
+		if type_ == "boolean" or type_ == "number" then
+            return
+        end
+		local count = counts[val]
+		counts[val] = (count or 0) + 1
+		if not count and type_ == "table" then
+			for k, v in pairs(val) do
+				count_values(k)
+				count_values(v)
 			end
-			nested[x] = nil
 		end
 	end
+	count_values(value)
+	return counts
+end
+
+-- Build a table with the succeeding character from A-Z
+-- Used to increment references quickly
+local succ = {}
+for char = ("A"):byte(), ("Z"):byte() - 1 do
+	succ[string.char(char)] = string.char(char + 1)
+end
 
-	local dumped     = {}  -- object->varname set
-	local local_defs = {}  -- Dumped local definitions as source code lines
+local function quote(string)
+	return string_format("%q", string)
+end
 
-	-- Mutually recursive local functions:
-	local dump_val, dump_or_ref_val
+local function dump_func(func)
+    return string_format("loadstring(%q)", string_dump(func))
+end
 
-	-- If x occurs multiple times, dump the local variable rather than
-	-- the value. If it's the first time it's dumped, also dump the
-	-- content in local_defs.
-	function dump_or_ref_val(x)
-		if seen[x] ~= 2 then
-			return dump_val(x)
+-- Serializes Lua nil, booleans, numbers, strings, tables and even functions
+-- Tables are referenced by reference, strings are referenced by value. Supports circular tables.
+local function serialize(value, write)
+    -- References are stored as tables of characters A-Z, which are digits of a base 26 number
+	local reference = {"A"}
+	local function increment_reference(place)
+		if not reference[place] then
+			reference[place] = "B"
+		elseif reference[place] == "Z" then
+            -- Carriage
+			reference[place] = "A"
+			return increment_reference(place + 1)
+		else
+			reference[place] = succ[reference[place]]
 		end
-		local var = dumped[x]
-		if var then  -- Already referenced
-			return var
+	end
```

Comment:
Indeed, this serves the purpose of shortening. `_` might have a special meaning, `0-9` are tricky because you might get something like `0x` which is tokenized as a number. I could include `a-z`, but that might be irritating to human readers...

Reply:
If you use all possible ASCII characters for names, you have of course make to sure that it's still valid names (ie. no numbers or keywords). This can be done by starting all names with an `A`.

I agree that including `a-z` can irritate human readers, especially with `l` and `I`. I hadn't thought about human-readability.

For simplicity I'd probably go with `"A"..counter`. But I don't really care.

Reply:
Alright then.